### DISCUSSION TITLE
Is Not Allowing The Compiling Of Unused Variables Really A Good Idea?

### DISCUSSION THREAD

**Discussion by Wajinn** [vlang/v#14193] [General]
In the documentation of V, it basically states that when dealing with unused variables, the V compiler will give a warning in development mode and not compile in production mode.

[V Docs Declaration-Errors Section](https://github.com/vlang/v/blob/master/doc/docs.md#declaration-errors)

However, this looks to be a questionable idea.  The reason why is that which variables are used can depend on data that is external to the program (like an external database or CSV file).  So (in pseudo code terms), it could be: `if data is string then assign to x, if data is integer then assign to y.`  

An argument could be made to use sum types and generics, but what if the variables are assigned to different subroutines/functions by their name, not their type?  They can't be differentiated by test of type, and the distinction between the variables is based upon the name (which is identified by the programmer or user action), and then they are acted upon by different subroutines/functions (for different purposes).

`if data is for Tom then assign to x (string), if data is for Jane then assign to y (string).`  For whom the data is for can't be known at compile-time, and can only be known at run-time.  The more obvious or convenient thing to do is to have a "blank" unused variable, that is assigned at run-time.

While it's understandable to give a warning about unused variables (to make sure the programmer isn't being sloppy), it does look like the production mode requirement is excessive and is boxing the programmer in.  Furthermore, this requirement doesn't appear to help with memory management, be it manual or automatic.

**Comment by medvednikov** [comment]
For this case there will be no warning :)

**Reply by Wajinn** [reply]
Thanks, I think, if I'm interpreting correctly.

**Reply by JalonSolov** [reply]
If you have any kind of reference to a variable, then is it not unused.  If you declare it then never do anything with it, it is unused.

In your example, the variable _is_ used, even if it's value isn't set until runtime.

Warning/error on unused var is a good thing... It may be because you refactored and don't use it any longer, declared 2 variables but only used one, etc., etc.  It's just sloppy to keep unused variables around, especially in production code.

**Reply by Wajinn** [reply]
Yes, I understand what you mean.  When compiling the program, it can appear that the variable is unused, because it was only declared and nothing else was done.  It can be the case that a variable appears unused at compile-time.  But at run-time, we want to use it.  If you heed the warning and remove the variable, then you can be removing possible wanted actions.  

Yes, it could be that the programmer simply has extraneous code that should be removed.  Maybe this code will be deemed extraneous and removed by the compiler.  Regardless, the compiler has done its job, with just the warning.  But, wouldn't it be better for the programmer to decide to leave it "as is" and proceed to compile with -prod?  At least, I don't see the harm.

Pseudo:
```
mut tom :=''
mut jane :=''

Default input equal blank.
Ask for user input, if none than do something else.
if get user input and ending is 'om', assign to tom.
if get user input and ending is 'ne', assign to jane.
```

**Reply by JalonSolov** [reply]
Do you do anything else with `tom` or `jane` (have to be lowercase for V) after that?  If not, then they are still unused, even if you assign something to them.

If you have other code that reads the values (regardless of where/when they were set), then they are NOT unused.

**Reply by Wajinn** [reply]
That would be a maybe, because it would depend on user input.  If no user input or incorrect input, the default is blank, so nothing done with tom and jane (per example).

I suppose one could write some extra code that doesn't really do anything.  So regardless of input or not, it would satisfy the compiler, there wouldn't be any warnings, and -prod could be used.  But, was curious about more direct ways.

**Reply by JalonSolov** [reply]
The whole point is to make you _use_ any variables you declare, otherwise you're wasting compile time, executable size, memory, etc., etc.

I mean if you're really _only_ assigning something, why bother?  If you don't read it again after assigning something to it, it's wasted space.

**Reply by Wajinn** [reply]
Yes, you are correct, that would be the ideal situation.  To use all variables declared.  But, I'm wondering if that can always be the case, when a program is to address unpredictable data or user input.

**Reply by JalonSolov** [reply]
Yes.  You either use the variable or not.  Doesn't matter how the value is set, whether it's fixed data, calculated data, or user input.

**Comment by Wajinn** [comment]
FYI. In the case of the compiler complaining of unused variables, it is possible to create a function to send such variables to. If this "dead" function has a return, but the function(s) that calls it doesn't assign the result to any variable, then it will satisfy the compiler so that -prod will work. Such actions doesn't appear to affect memory management (as is), particularly when automatic, though clearly this is not common nor good practice.

Intriguingly, though a bit unrelated, all variables can be sent to such a "dead" function or the functionality of such a function can be changed for some creative, useful, or monitoring purposes.

**Reply by JalonSolov** [reply]
:man_shrugging: You can also do pretty much anything in an `unsafe{}` block, but that isn't recommended either.  :-)
